Initial ostree support

gicmo opened the pull request:
This adds support for creating ostree commits and image based on ostree commits. For this it adds:

- **`sources`**:
   - `org.osbuild.ostree`: can obtain ostree commits and caches them in a local repo
- **`stages`**:
  - `org.osbuild.rpm-ostree`: does postprocessing on a "final" file system tree to adapt it for ostree
  - `org.osbuild.ostree`: bootstraps a filesystem tree, in ostree layout, for a given commit
  - `org.osbuild.ostree.grub2`: install grbu2 in a ostree installation
- **`assemblers`**:
  - `org.osbuild.ostree.commit`: take a ostree filesystem and commit it; result is a repository with it

Adds some sample files showing off how it fits together. This closes #260

gicmo commented:
Essential things missing here are:
  - [ ] pylint errors
  - [ ]  cleanup `STAGE_OPTS`, `STAGE_DESC`, `STAGE_INFO`
      - [ ] schema description for the stages

teg commented on assemblers/org.osbuild.ostree.commit:
```diff
@@ -0,0 +1,114 @@
+        print("Initializing root filesystem", file=sys.stderr)
+        init_rootfs(root, tmp_is_dir)
+
+        print("Copying data", file=sys.stderr)
```
This we would ideally move, right? If we had been passed in the parent of `tree`?

gicmo replied:
Yes, that is the plan still. The whole file system cleanup should be moved to the `org.ostree.rpm-ostree` stage (or a completely new one).

teg commented on assemblers/org.osbuild.ostree.commit:
```diff
@@ -0,0 +1,114 @@
+        if parent:
+            argv += [f"--parent={parent}"]
+
+        argv += [f"--write-commitid-to={output_dir}/commit.id",
```
Was there the possibility of outputting a tar file instead of a directory? What would be the pros/cons of that? To me seems nice to have a single artefact, but I can imagen there are considerations at play.

gicmo replied:
I would wait until we have assembler pipelines to produce the tar file with an tar assembler stage.

teg commented on samples/f30-ostree.json:
```diff
@@ -0,0 +1,109 @@
+{
+  "sources": {
+    "org.osbuild.dnf": {
```
This is easier said than done, perhaps, but I think we should move away from using the dnf stage anywhere, so that we can drop it completely soon.

gicmo replied:
Absolutely. Done, by actually ;)

gicmo commented:
Quick update: This is almost good to be reviewed. 

The major SELinux issues have been fixed, especially `/etc` and `/boot` are labelled corrected; `/sysroot` and ironically `/var/ostree-selabeled ` still have wrong labels. For `/sysroot` there is no entry in the `file_contexts` file, so I assume it would normally inherit the one from `/` but not in our case because there we build in a temp dir. Might have to file a bug upstream and work around it for now.

The other big thing I want to do is using the normal `grub2` stage.

gicmo commented:
Latest update: added support for ostree to the normal grub2 stage and dropped the special one. Since we need to copy the EFI binary data I had to add `/boot` to the build root mount points. With this even hybrid boot support works.

There is still the issue of the wrongly labeled `/sysroot`, I might have to apply a work around for that, but that should not block the review of this.

teg commented on sources/org.osbuild.ostree:
```diff
@@ -30,13 +31,22 @@ def main(options, checksums, cache, output):
         uid = remote["uid"]
+        gpg = remote.get("gpgkeys", [])
+
+        extra_args = []
```
In the `org.osbuild.rpm` stage we do the gpg verification in the stage, rather than in the source. We should probably unify here. I was thinking the gpg keys would be a property of the given manifest, and that they should be rechecked if the next manifest uses different keys, but the data is still available in the source cache, but maybe it makes sense to verify on import?

gicmo replied:
Good point. This somehow felt natural to me and I forgot to check what the rpm stage does. I felt that once it is in the cache it is trusted but maybe re-checking is also a good idea? I can argue both sides in my head and they both seem to make sense one way or the other.

teg replied:
Yeah, I can see both sides too... Something to discuss (and document as a principle, I guess).

teg commented on stages/org.osbuild.selinux:
```diff
@@ -38,21 +45,55 @@ STAGE_OPTS = """
+    else:
+        file_contexts = os.path.join(tree, file_contexts)
+
+    setfiles(tree, paths, file_contexts)
```
Should this `setfiles()` be in the conditional?

In general for selinux: if we do the labelling before creating the commit, is it not preserved? Does it end up wrong?

gicmo replied:
This is intended since I am labelling the global `/boot` with it (but using the context file from the deployment) but also the `/etc` that is within the deployment:

```js
        "options": {
          "file_contexts": "etc/selinux/targeted/contexts/files/file_contexts",
          "paths": [
            "/boot"
          ],
          "ostree": {
            "commit": "711cb14fc153902b2aa75ca26a88f51269eebf4d1763bb33617b9963c6045f59",
            "osname": "fedora",
            "paths": [
              "/etc"
            ]
          }
        }
```

In general: we do level before the commit, in the `ostree.commit` stage. But when creating the image, we do a `ostree deploy`, which creates new files (three way merge of `/etc`, first and foremost, but also `/boot` because the kernel+initramfs+bls snippets get installed). Normally ostree would label those, but that internally depends on `is_selinux_enabled` being `TRUE`, which in our container it is not, we won't get proper labels. Hence we have to label in both pipelines.

teg replied:
Hm, could we consider this a bug in `ostree deploy` and just do the logic they were meant to unconditionally there? Just to get the API surface "correct"?

gicmo replied:
So I did that emulate what `ostree deploy` is doing in the stage and was happy until I realized that the `grub2` stage comes afterwards and that also 1) copies files and 2) creates new ones in `/boot/grub2/` (`grub.cfg`), 3) in `boot/efi/EFI/<vendor>/grub.cfg` as well as 4) `/etc` (`grubenv`). For 1) correctly labelling them in the build root could and then relying on `shutil.copy2` to copy the labels as well could work. For 2) we need to install `grub2-pc` in the build host copy that file over and modify it. For 3) we are out of luck, but it will end up on a FAT partition, so it might be fine. 4) is a real issue in two ways: it is in created in the totally wrong place (it should be in the deployroot) and then obviously has needs to be relabelled there.

gicmo replied:
One idea to solve this could be to split the `org.osbuild.ostree` stage into two parts. the one that does preparses the filesystem (`ostree admin init-fs`) and then pulls (`ostree admin pull-local`) the commit (`org.osbuild.ostree.init` or `pull` ) and another one that deploys the commit (`ostree admin os-init` and `ostree admin deploy`) and have the `grub2` stage run between the two. Since in that step we fix labels in `/boot` and `deployroot/etc` this would "accidentally :("  also fix the labels of the grub2 stage. Or we keep the changes to selinux stage and run that afterwards. Or maybe better, create a new `org.osbuild.ostree.selinux` stage that does the fixups.

gicmo replied:
Ok, ignore that idea, because I need to of course have the tree deployed for a `deployroot` to exist, so the `grub2` stage needs to come after the the `deploy`stage. :chicken: :egg:

gicmo replied:
After various discussion, including with @martinezjavier, I introduced a new boolean, which controls whether the `/etc/default/grub` file is written at all. The idea being that no automatic tools are actually using it (it used to be grubby but that is deprecated) so it is only read if the user actively does `grub2-mkconfig -o ...`. I am not convinced that in a BLS scenario anyone would really want to run that command, instead of maybe editing `grub.cfg` directy (to tweak the timeout). Anyway, @martinezjavier and I agreed that  `/etc/grub.d/10_linux` will could be changed so that `GRUB_ENABLE_BLSCFG=true` is the default if `grubby-deprecated` is not installed (i.e. `/usr/sbin/new-kernel-pkg` not present). Then we would not need the file at all anymore, even for `grub2-mkconfig`. The only other thing we write is `GRUB_TIMEOUT` which would be reset to the default, which can easily be adjust be the user, who apparently knows what they are doing. We also don't define a bunch of other variables, so the resulting `grub.cfg` would differ from the one we wrote anyway.

teg commented on stages/org.osbuild.grub2:
```diff
@@ -137,6 +151,7 @@ def main(tree, options):
     kernel_opts = options.get("kernel_opts", "")
     legacy = options.get("legacy", True)
     uefi = options.get("uefi", None)
+    ostree = options.get("ostree", False)
```
Is this the right name for the option? Is it really ostree specific?

gicmo replied:
You are right, this evolved from using `ostree.{commit, osname}` to locate the efi binaries of the deployment into using the efi binaries from the build root, and now could just be `copy_efi_binaries` or something.

gicmo replied:
I went with `uefi.install` in the end.

gicmo replied:
One problem: the `grubenv` file gets created in the wrong place: it will be created in `<diskroot>/etc/grubenv` not in `/ostree/deploy/<stateroot>/deploy/<commit>.0/etc`. So maybe the `grub2` stage needs to have a `ostree` option after all.

gicmo replied:
Not `/etc/grubenv` but `/etc/default/grub`, so the discussion at the selinux stage.

dvdhrm commented on osbuild/ostree.py:
```diff
@@ -0,0 +1,90 @@
+
+
+class Param:
+    def __init__(self, value_type, mandatory=False, default=None):
```
How about:
`def __init__(self, value_type, *, mandatory=False, default=None):`

This forces call-sites to use named arguments, which makes them a lot nicer to read than just `true` or `false`.

dvdhrm replied:
We discussed this separately. To make sure we have this documented: The unannotated star `*` does exactly what we expect: It forces the following parameters to be keyword-arguments. It has no other side-effects. It does not swallow other positional arguments.

We can adjust this later on as well. But I think forced keyword-arguments seems like a nice feature.

gicmo replied:
I am on the fence here because I realized I wanted no a 'real' keyword argument but an normal optional argument with a default value. But I don't feel strongly about it here.

gicmo commented on osbuild/ostree.py:
```diff
@@ -0,0 +1,90 @@
+    def __init__(self):
+        self._data = {}
+        self["ref"] = "osbuild/devel"
+        self["repos"] = ["osbuild"]
```
`repos` we don't actually use, it is used in `rpm-ostree compose` for downloading packages and we don't use that part, BUT the `Treefile` says it is a required element and I don't want to upset the parser. Similar situation for `ref`, only used by us in the commit stage, but technically required to be present in a valid `Treefile`.

gicmo commented on osbuild/ostree.py:
```diff
@@ -52,8 +52,12 @@ class Treefile:
+        self["boot-location"] = "new"
+        self["tmp-is-dir"] = True
+        self["ignore-removed-users"] = ["root"]
+        self["ignore-removed-groups"] = ["root"]
```
Ah, no these should go, was an oversight. Thanks

gicmo commented on stages/org.osbuild.rpm-ostree:
```diff
@@ -0,0 +1,77 @@
+    "type": "array",
+    "items": { "type": "string" }
+  },
+  "machineid_compat": {
```
We have a choice here, either [`rpm-ostree`](https://github.com/coreos/rpm-ostree/blob/1cf0d557ae8059e689b1fed670022727e9842288/src/libpriv/rpmostree-postprocess.c#L396) will write an empty one a it will be removed. Once, in the future, we roll our own code that does what `rpm-ostree compose postprocess` does, we don't have to care anymore but currently we need to tell it something. 

I agree that it should not be a option though. Read the machine id and reset it after `rpm-ostree` has run?

teg replied:
Sounds good to me, if this is the intended behaviour in the end.

gicmo replied:
This is changed so we detect an existing file and adjust the setting accordingly.

dvdhrm commented on osbuild/ostree.py:
```diff
@@ -0,0 +1,90 @@
+            fd, name = tempfile.mkstemp(suffix=".json",
+                                        text=True)
+
+            with os.fdopen(fd, "w+") as f:
```
Does `fdopen()` close the fd even if it raises an exception for other reasons? The docs are not really clear, and I got lost in the c-sources. But we also probably don't really care, as we re-raise the exception anyway.

gicmo replied:
Good that you asked, I was wondering exactly the same when I wrote the code but then forgot about it again. Sadly that means I also don't know ;)

dvdhrm replied:
Hehe, I see. I checked the sources and you can easily make `io.open()` trigger an exception without closing the passed file-descriptor [1]. But you can also cause it to close it [2]. But I think as long as the arguments are valid, it will always end up closing it under `OSError`s. Regardless, nothing we can do about it, other than assuming it cleaned it up correctly.

[1] https://github.com/python/cpython/blob/master/Lib/_pyio.py#L170
[2] https://github.com/python/cpython/blob/master/Lib/_pyio.py#L256

gicmo commented:
> The commit-ordering seems off, but I blame github.

Me too!

> Otherwise, I would make the gpg-verification part of `pull-local`, which then enforces a verification on every checkout. Whether you want to keep it on the remote pull is up to you. I would drop it, but feel free to keep it.

Done.

> Looks good to me. Still room for improvements, but nothing that should stop this from getting merged. Thanks a lot for this huge effort!

:partying_face:
